internal/controller/siteconfig_builder.go

```diff
@@ -37,16 +40,113 @@ func NewSiteConfigBuilder(pLog logr.Logger) *SiteConfigBuilder {
 	return &scBuilder
 }
 
-// getConfigMap retrieves the configmap from cluster for a given TemplateRef
-func getConfigMap(ctx context.Context, c client.Client, templateRef v1alpha1.TemplateRef) (*corev1.ConfigMap, error) {
+func (scbuilder *SiteConfigBuilder) ProcessTemplates(ctx context.Context, c client.Client, siteConfig *v1alpha1.SiteConfig) ([]interface{}, error) {
+	scbuilder.Log.Info(fmt.Sprintf("Processing cluster-level templates for SiteConfig %s", siteConfig.Name))
 
-	cm := &corev1.ConfigMap{}
-	if err := c.Get(ctx, types.NamespacedName{
-		Name:      templateRef.Name,
-		Namespace: templateRef.Namespace,
-	}, cm); err != nil {
+	// Render cluster-level templates
+	siteManifests, err := scbuilder.renderTemplates(ctx, c, siteConfig, siteConfig.Spec.TemplateRefs, -1, siteConfig.Spec.SuppressedManifests)
+	if err != nil {
+		scbuilder.Log.Info(fmt.Sprintf("encountered error while processing cluster-level templates for SiteConfig %s, err: %s", siteConfig.Name, err.Error()))
+		return siteManifests, err
+	}
+	scbuilder.Log.Info(fmt.Sprintf("Processed cluster-level templates for SiteConfig %s", siteConfig.Name))
+
+	// Process node-level templates
+	numNodes := len(siteConfig.Spec.Nodes)
+	for nodeId, node := range siteConfig.Spec.Nodes {
+		scbuilder.Log.Info(fmt.Sprintf("Processing node-level templates for SiteConfig %s [node: %d of %d]", siteConfig.Name, nodeId+1, numNodes))
+
+		// Render node-level templates
+		nodeManifests, err := scbuilder.renderTemplates(ctx, c, siteConfig, node.TemplateRefs, nodeId, siteConfig.Spec.Nodes[nodeId].SuppressedManifests)
```

Comment:
Potential bug: you are passing the address (*v1alpha1.SiteConfig) of a loop pointer ref `node.TemplateRefs`

Reply:
Good catch!